php/libraries/SinglePointLogin.class.inc

```diff
@@ -308,85 +308,96 @@ class SinglePointLogin
+            }
 
-                if ($row['Pending_approval'] == 'Y') {
-                    $this->_lastError = "Your account has not yet been activated."
-                        . " Please contact your project administrator to activate"
-                        . " this account.";
-                    $this->insertFailedDetail("user account pending", $setArray);
-                    return false;
-                }
+            // Check whether account needs to be approved by an admin
+            if ($row['Pending_approval'] == 'Y') {
+                $this->_lastError = "Your account has not yet been activated."
+                    . " Please contact your project administrator to activate"
+                    . " this account.";
+                $this->insertFailedDetail("user account pending", $setArray);
+                return false;
+            }
+            if ($row['Active'] == 'N'
+                || $this->disabledDueToInactivity($username, 365)
+            ) {
+                $this->_lastError = "Your account has been deactivated."
+                    . " Please contact your project administrator to"
+                    . " reactivate this account.";
+                $this->insertFailedDetail(
+                    "user account not active",
+                    $setArray
+                );
+
+                return false;
+            }
 
-                // save the new password if the last password expired
-                if ($row['Expired']) {
-                    $success = $this->save();
+            // Save the new password if the last password expired
+            if ($row['Expired']) {
+                $success = $this->save();
 
-                    if ($success == false) {
-                        if ($redirect) {
-                            $this->showPasswordExpiryScreen();
-                        }
-                        return false;
+                if ($success == false) {
+                    if ($redirect) {
+                        $this->showPasswordExpiryScreen();
                     }
+                    return false;
                 }
-
-                // user is valid
-                $this->_username = $row['UserID'];
-
-                $DB->insert('user_login_history', $setArray);
-                return true;
             }
-        }
+            // Allow a user to login if none of the previous cases have occurred
+            $this->_username = $row['UserID'];
 
-        // bad usename or password
-        $this->_lastError = "Incorrect email or password";
+            $DB->insert('user_login_history', $setArray);
+            return true;
+        }
+        /* When the password is wrong, log this detail for rate-limiting
+         * purposes.
+         */
+        $this->insertFailedDetail(
+            "Bad password",
+            $setArray
+        );
+
+        // Fail by default with a generic error message
+        $this->_lastError = "Incorrect username or password.";
```

Comment:
_lastError should already be set to this value from line 316 no?

Reply:
In theory yes and I removed this in an earlier version of this PR. @driusan had me add it back so we have this set up to fail by default unless the user meets all of the above criteria.

I agree with you that the logic is confusing to read.